Configurable file watcher debounce duration in live-reload mode

ccjmne opened the pull request:
- Introduce `-d` `--debounce <VALUE>` flag for `serve`
- Document `serve`'s `--debounce` flag in `cli-usage.md`

Let the file watcher use a configurable debounce value.

This commit preserves the (until now hard-coded) 1-second debounce value
as the default, but adds a new flag for the `serve` command, which lets
the user configure a debounce duration for consecutive rebuilds in watch
mode.

Closes #2937.

ccjmne committed:
Introduce -d|--debounce <VALUE> flag for serve

Let the file watcher use a configurable debounce value.

This commit preserves the (until now hard-coded) 1-second debounce value
as the default, but adds a new flag for the `serve` command, which lets
the user configure a debounce duration for consecutive rebuilds in watch
mode.

Closes #2937.

Ref: https://github.com/getzola/zola/issues/2937
Ref: https://zola.discourse.group/t/watch-debounce-configuration/2666

ccjmne committed:
Document serve's --debounce flag in cli-usage.md

Ref: 6737349447f79314c15fdd860b26eeb899f38d86
Ref: https://github.com/getzola/zola/issues/2937
Ref: https://zola.discourse.group/t/watch-debounce-configuration/2666

nicolandu commented:
Also, it would be good to document the flag (`-d`/`--debounce`) more thoroughly, such as by specifying its unit (milliseconds) in the docs.

ccjmne committed:
Explicitly prohibit --debounce 0

The `tick_rate` at which the debouncer library checks whether any events
should now be dispatched shouldn't be 0.  It cannot either be any
greater than the debounce rate itself.

Ideally, we may want to *disable* debouncing altogether when the user
passes --debounce 0, but that'd require a special case to be made in the
instrumentation, where we'd bypass the debouncer: that's another package
to depend on, and ultimately deemed no worth the hassle.

Ref: https://github.com/getzola/zola/issues/2937
Ref: https://zola.discourse.group/t/watch-debounce-configuration/2666

ccjmne commented:
Done, and done.

I looked into the reason for the excessive CPU load with `--debounce 0`, and it seems to be a limitation of the debouncing library: it's not made to work efficiently with `0` there—which is reasonable.

The "ideal" resolution may have been to somehow bypass using the debouncer altogether if the user passes `--debounce 0`, but frankly, having another dependency (the debouncing library doesn't expose its underlying "watcher)" and the extra code to handle that special case seems like too much of a maintenance burden to pay for this little of a nitpick.

My puny mammalian brain can't really feel the difference between 1ms and 0ms anyways ;)

Keats commented:
The code looks fine, i'm just a bit hesitant about adding more flags to the CLI

ccjmne committed:
Fix formatting style

The line was too long.  I pruned the /*tick_rate=*/ inline comment that
was indicating the name of the argument it was annotating, since it was
*the exception*: there are zero other occurrences of such notation
throughout this codebase.

ccjmne commented:
> No but with auto-saving editors it might happen frequently enough

Oh alright. I would presume then that the debouncing happens at *that* level, where the editor wouldn't save your file to disk after each keystroke.

For myself, waiting one full second after I've virtually instructed Zola to rebuild my site is a bit of an irritation.
It's probably vastly exacerbated by the fact I'm still in the "build up the theme" phase, where I tweak styles and want to iterate rapidly: I can see how authoring *content* needs much less reactivity :)

nicolandu commented:
I just had a thought: what if we lower the debounce from 1000 ms to a lower value, say 50 or 100 ms, and ditch the complexity of an extra CLI flag?

ccjmne commented:
@Keats, I think we'd all, but you primarily, benefit from understanding what is going on.

It's quite possible that you've configured something to commit changes, as you type, back to your filesystem every 100ms, apparently with some throttle: naively, I'd say that's an undesirable configuration.

Regardless, I truly don't think that your specific, undiagnosed set-up being quite curious should be the sole blocker that prevents a 3ms rebuild to take anything less than 1003ms.

--

Here's the isolated, minimal, complete, verifiable example I could set up. You only need Docker:

1. Start with a fresh, empty distribution, install cargo, git, and some text editor, compile zola with the `--debounce` flag, serve the test site:

```sh
docker run --rm -it --network=host --name=zola archlinux sh
pacman -Sy --noconfirm git vim rust cargo
git clone --depth 1 --branch configurable-watcher-debounce https://github.com/ccjmne/zola
cd zola; cargo build --release
cd test-site; ../target/release/zola serve -fd1
```
On your local machine, navigate to `localhost:1111`.

2. Inside the Docker container, edit something like `templates/index.html` and see it work splendidly, just taking a handful of milliseconds to rebuild and republish each time you purposely write your file:

```sh
docker exec -it zola vim /zola/test_site/templates/index.html
```

--

Zola truly is a pretty neat tool, thank you for that!

Now if you still think that, in principle, waiting around is what is ultimately desirable for the project, that's entire acceptable. If you truly suspect that *any* (let alone a very large) debouncing is required, I'm not entirely convinced, and can't quite justify that in my (admittedly modest) test.
